network/discovery/discovery.go

```diff
@@ -0,0 +1,427 @@
+	GetRandomBootnode() *peer.AddrInfo
+
+	// GetBootnodeConnCount fetches the number of bootnode connections [Thread safe]
+	GetBootnodeConnCount() int64
+
+	// PROTOCOL MANIPULATION //
+
+	// NewDiscoveryClient returns a discovery gRPC client connection
+	NewDiscoveryClient(peerID peer.ID) (proto.DiscoveryClient, error)
+
+	// CloseProtocolStream closes a protocol stream to the peer
+	CloseProtocolStream(protocol string, peerID peer.ID) error
+
+	// PEER MANIPULATION //
+
+	// DisconnectFromPeer attempts to disconnect from the specified peer
+	DisconnectFromPeer(peerID peer.ID, reason string)
+
+	// AddToPeerStore adds a peer to the networking server's peer store
+	AddToPeerStore(peerInfo *peer.AddrInfo)
+
+	// GetPeerInfo fetches the peer information from the server's peer store
+	GetPeerInfo(peerID peer.ID) *peer.AddrInfo
+
+	// GetRandomPeer fetches a random peer from the server's peer store
+	GetRandomPeer() *peer.ID
+
+	// TEMPORARY DIALING //
+
+	// FetchAndSetTemporaryDial checks if the peer connection is a temporary dial,
+	// and sets a new value accordingly
+	FetchOrSetTemporaryDial(peerID peer.ID, newValue bool) bool
+
+	// RemoveTemporaryDial removes a peer from the temporary dial map
+	RemoveTemporaryDial(peerID peer.ID)
+
+	// CONNECTION INFORMATION //
+
+	// HasFreeConnectionSlot checks if there is an available connection slot for the set direction [Thread safe]
+	HasFreeConnectionSlot(direction network.Direction) bool
+}
+
+// DiscoveryService is a service that finds other peers in the network
+// and connects them to the current running node
+type DiscoveryService struct {
+	proto.UnimplementedDiscoveryServer
+
+	baseServer   networkingServer // The interface towards the base networking server
+	logger       hclog.Logger     // The DiscoveryService logger
+	routingTable *kb.RoutingTable // Kademlia 'k-bucket' routing table that contains connected nodes info
+
+	closeCh chan struct{} // Channel used for stopping the DiscoveryService
+}
+
+// NewDiscoveryService creates a new instance of the discovery service
+func NewDiscoveryService(
+	server networkingServer,
+	routingTable *kb.RoutingTable,
+	logger hclog.Logger,
+	closeCh chan struct{},
+) *DiscoveryService {
+	return &DiscoveryService{
+		logger:       logger.Named("discovery"),
+		baseServer:   server,
+		routingTable: routingTable,
+		closeCh:      closeCh,
+	}
+}
+
+// Start starts the discovery service
+func (d *DiscoveryService) Start() {
+	go d.startDiscovery()
+}
+
+// Close stops the discovery service
+func (d *DiscoveryService) Close() {
+	close(d.closeCh)
+}
+
+// RoutingTableSize returns the size of the routing table
+func (d *DiscoveryService) RoutingTableSize() int {
+	if d.routingTable == nil {
+		return 0
+	}
+
+	return d.routingTable.Size()
+}
+
+// RoutingTablePeers fetches the peers from the routing table
+func (d *DiscoveryService) RoutingTablePeers() []peer.ID {
+	if d.routingTable == nil {
+		return []peer.ID{}
+	}
```

Comment:
Is there a possibility of that d.routingTable is nil?

Reply:
Good catch 💯 

This was leftover. There is no possibility for the routing table to **not** be initialized.

Fixed:

[cdda743](https://github.com/0xPolygon/polygon-edge/pull/343/commits/cdda7432b5f1e21c83e592cb72d6d72ccae46229)